lib/server-builder.js

```diff
@@ -0,0 +1,154 @@
+import { SubProcess } from 'teen_process';
+import { fs, logger } from 'appium-support';
+import log from './logger';
+import B from 'bluebird';
+import _ from 'lodash';
+import path from 'path';
+import { EOL } from 'os';
+
+const IGNORED_ERRORS = [];
+const GRADLE_VERSION_KEY = 'gradle_version';
+const GRADLE_URL_PREFIX = 'distributionUrl=';
+const GRADLE_URL_TEMPLATE = 'https\\://services.gradle.org/distributions/gradle-VERSION-all.zip';
+const GRADLE_MAX_ERROR_LOG_LINES = 15;
+
+const VERSION_KEYS = [
+  GRADLE_VERSION_KEY,
+  'agp_version',
+  'compile_sdk_version',
+  'build_tools_version',
+  'min_sdk_version',
+  'target_sdk_version',
+  'kotlin_version'
+];
+
+const gradleLog = logger.getLogger('Gradle');
+
+class ServerBuilder {
+  server_versions
+
+  constructor (args = {}) {
+    this.serverPath = args.serverPath;
+    this.showGradleLog = args.showGradleLog;
+
+    const versions_arg = args.versions || {};
+    const server_versions = {};
+    for (const key of VERSION_KEYS) {
+      if (versions_arg[key]) {
+        server_versions[key] = versions_arg[key];
+      }
+    }
+    this.server_versions = server_versions;
+  }
+
+  async build () {
+    if (this.server_versions[GRADLE_VERSION_KEY]) {
+      await this.setGradleWrapperVersion(this.server_versions[GRADLE_VERSION_KEY]);
+    }
+    await this.runBuildProcess();
+  }
+
+  getCommand () {
+    let cmd = `.${path.sep}gradlew`;
+    let args = [];
+
+    for (const key of VERSION_KEYS) {
+      const serverVersion = this.server_versions[key];
+      if (key !== GRADLE_VERSION_KEY && serverVersion) {
+        args.push(`-P${key}=${serverVersion}`);
+      }
+    }
+
+    args.push('assembleAndroidTest');
+
+    return {cmd, args};
+  }
+
+  async setGradleWrapperVersion (version) {
+    const propertiesPath = path.resolve(this.serverPath, 'gradle', 'wrapper', 'gradle-wrapper.properties');
+    const originalProperties = await fs.readFile(propertiesPath, 'utf8');
+    const newProperties = this.updateGradleDistUrl(originalProperties, version);
+    await fs.writeFile(propertiesPath, newProperties, 'utf8');
+  }
+
+  updateGradleDistUrl (propertiesContent, version) {
+    return propertiesContent.split('\n')
+      .map(function (line) {
+        if (line.startsWith(GRADLE_URL_PREFIX)) {
+          return GRADLE_URL_PREFIX + GRADLE_URL_TEMPLATE.replace('VERSION', version);
+        } else {
+          return line;
+        }
+      })
+      .join('\n');
+  }
+
+  runBuildProcess () {
+    const {cmd, args} = this.getCommand();
+    log.debug(`Beginning build with command '${cmd} ${args.join(' ')}' ` +
+      `in directory '${this.serverPath}'`);
+    const gradlebuild = new SubProcess(cmd, args, {
+      cwd: this.serverPath,
+      env: process.env,
+      detached: true,
+      stdio: ['ignore', 'pipe', 'pipe'],
+    });
+    gradlebuild._last_lines = [];
```

Comment:
`lastLines`

Reply:
why should this variable be attached to gradlebuild ?

Reply:
Here I've just copied approach from xcuitest driver. I don't think there are any issues with having two separate variables.